File: ios-sdui/Sources/HSBC_SDUI/Components/KYC/KYCComponents.swift
import Foundation
import SwiftUI
import Combine
// HIVE tokens imported from hive-tokens/swift/HiveTokens.swift
// Copy HiveTokens.swift into this Swift Package / Xcode target before building.

// ─── KYC Models ──────────────────────────────────────────────────────────────

struct KYCValidation: Codable {
    let required: Bool?
    let minLength: Int?
    let maxLength: Int?
    let pattern: String?
    let errorMessage: String?
    let minAge: Int?
}

struct KYCSelectOption: Codable {
    let value: String
    let label: String
}

// ─── KYC Text Input (iOS) ────────────────────────────────────────────────────

struct KYCTextInputView: View {
    let questionId: String
    let label: String
    let placeholder: String?
    let helpText: String?
    let validation: KYCValidation?
    let savedValue: String?
    let onAnswer: (String, String) -> Void

    @State private var value: String
    @State private var error: String?
    @State private var touched = false

    init(questionId: String, label: String, placeholder: String? = nil,
         helpText: String? = nil, validation: KYCValidation? = nil,
         savedValue: String? = nil, onAnswer: @escaping (String, String) -> Void) {
        self.questionId = questionId
        self.label = label
        self.placeholder = placeholder
        self.helpText = helpText
        self.validation = validation
        self.savedValue = savedValue
        self.onAnswer = onAnswer
        _value = State(initialValue: savedValue ?? "")
    }

    var body: some View {
        VStack(alignment: .leading, spacing: HiveSpacing.s1) {
            HStack(spacing: HiveSpacing.s1) {
                Text(label)
                    .font(HiveTypography.labelBase)
                    .foregroundColor(HiveColor.neutral.n700)
                if validation?.required == true {
                    Text("*").foregroundColor(HiveColor.semantic.error)
                }
            }
            if let help = helpText {
                Text(help)
                    .font(HiveTypography.caption)
                    .foregroundColor(HiveColor.neutral.n500)
            }
            TextField(placeholder ?? "", text: $value)
                .font(HiveTypography.bodyBase)
                .padding(.horizontal, HiveComponent.input.paddingH)
                .frame(height: HiveComponent.input.height)
                .background(HiveComponent.input.bgColor)
                .cornerRadius(HiveComponent.input.borderRadius)
                .overlay(
                    RoundedRectangle(cornerRadius: HiveComponent.input.borderRadius)
                        .stroke(
                            touched && error != nil
                                ? HiveComponent.input.errorColor
                                : HiveComponent.input.borderColor,
                            lineWidth: 1
                        )
                )
                .autocapitalization(.words)
                .onSubmit { validate(); onAnswer(questionId, value) }
                .onChange(of: value) { _ in if touched { validate() } }
            if touched, let err = error {
                Text(err)
                    .font(HiveTypography.caption)
                    .foregroundColor(HiveColor.semantic.error)
            }
        }
        .padding(.vertical, HiveSpacing.s1)
    }

    private func validate() {
        touched = true
        guard let v = validation else { error = nil; return }
        if v.required == true && value.trimmingCharacters(in: .whitespaces).isEmpty {
            error = v.errorMessage ?? "This field is required"; return
        }
        if let min = v.minLength, value.count < min {
            error = "Minimum \(min) characters"; return
        }
        if let pattern = v.pattern,
           !NSPredicate(format: "SELF MATCHES %@", pattern).evaluate(with: value) {
            error = v.errorMessage ?? "Invalid format"; return
        }
        error = nil
    }
}

// ─── KYC Single Select — Bottom Sheet (iOS native) ────────────────────────────

struct KYCSingleSelectView: View {
    let questionId: String
    let label: String
    let options: [KYCSelectOption]
    let savedValue: String?
    let onAnswer: (String, String) -> Void

    @State private var selected: String?
    @State private var showSheet = false

    init(questionId: String, label: String, options: [KYCSelectOption],
         savedValue: String? = nil, onAnswer: @escaping (String, String) -> Void) {
        self.questionId = questionId; self.label = label
        self.options = options; self.savedValue = savedValue; self.onAnswer = onAnswer
        _selected = State(initialValue: savedValue)
    }

    var selectedLabel: String {
        options.first { $0.value == selected }?.label ?? "Select an option"
    }

    var body: some View {
        VStack(alignment: .leading, spacing: HiveSpacing.s1) {
            Text(label)
                .font(HiveTypography.labelBase)
                .foregroundColor(HiveColor.neutral.n700)
            Spacer().frame(height: HiveSpacing.s1)
            Button(action: { showSheet = true }) {
                HStack {
                    Text(selectedLabel)
                        .foregroundColor(selected == nil ? HiveColor.neutral.n400 : HiveColor.neutral.n800)
                        .font(HiveTypography.bodyBase)
                    Spacer()
                    Image(systemName: "chevron.down")
                        .foregroundColor(HiveColor.neutral.n500)
                        .font(.system(size: 12))
                }
                .padding(.horizontal, HiveComponent.input.paddingH)
                .frame(height: HiveComponent.input.height)
                .background(HiveComponent.input.bgColor)
                .cornerRadius(HiveComponent.input.borderRadius)
                .overlay(
                    RoundedRectangle(cornerRadius: HiveComponent.input.borderRadius)
                        .stroke(HiveComponent.input.borderColor, lineWidth: 1)
                )
            }
        }
        .padding(.vertical, HiveSpacing.s1)
        .confirmationDialog(label, isPresented: $showSheet, titleVisibility: .visible) {
            ForEach(options, id: \.value) { option in
                Button(option.label) {
                    selected = option.value
                    onAnswer(questionId, option.value)
                }
            }
            Button("Cancel", role: .cancel) {}
        }
    }
}

// ─── KYC Progress Bar (iOS) ──────────────────────────────────────────────────

struct KYCProgressBarView: View {
    let currentStep: Int
    let totalSteps: Int
    let sectionTitle: String
    let sectionProgress: String?

    var progress: Double { Double(currentStep) / Double(totalSteps) }

    var body: some View {
        VStack(alignment: .leading, spacing: HiveSpacing.s1) {
            HStack {
                Text(sectionTitle)
                    .font(HiveTypography.caption)
                    .foregroundColor(HiveColor.neutral.n500)
                Spacer()
                Text(sectionProgress ?? "Step \(currentStep) of \(totalSteps)")
                    .font(HiveTypography.caption)
                    .foregroundColor(HiveColor.neutral.n500)
            }
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    Capsule()
                        .fill(HiveComponent.progressBar.trackColor)
                        .frame(height: HiveComponent.progressBar.height)
                    Capsule()
                        .fill(HiveComponent.progressBar.fillColor)
                        .frame(width: geo.size.width * progress,
                               height: HiveComponent.progressBar.height)
                        .animation(HiveAnimation.slow, value: progress)
                }
            }
            .frame(height: HiveComponent.progressBar.height)
        }
        .padding(.horizontal, HiveSpacing.s4)
        .padding(.vertical, HiveSpacing.s2)
    }
}

// ─── KYC Document Upload (iOS — camera first) ────────────────────────────────

struct KYCDocUploadView: View {
    let questionId: String
    let label: String
    let helpText: String?
    let onAnswer: (String, Data) -> Void

    @State private var capturedImage: UIImage?
    @State private var showImagePicker = false
    @State private var sourceType: UIImagePickerController.SourceType = .camera

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(label).font(.subheadline).fontWeight(.semibold)
            if let help = helpText {
                Text(help).font(.caption).foregroundColor(.secondary)
            }
            if let img = capturedImage {
                Image(uiImage: img)
                    .resizable().scaledToFit()
                    .frame(maxHeight: 200)
                    .cornerRadius(8)
                Button("Retake Photo") {
                    capturedImage = nil
                    sourceType = .camera
                    showImagePicker = true
                }
                .foregroundColor(.red)
            } else {
                VStack(spacing: 12) {
                    Button(action: { sourceType = .camera; showImagePicker = true }) {
                        Label("Take Photo", systemImage: "camera")
                            .frame(maxWidth: .infinity).padding(14)
                            .background(Color.red).foregroundColor(.white)
                            .cornerRadius(8)
                    }
                    Button(action: { sourceType = .photoLibrary; showImagePicker = true }) {
                        Label("Choose from Gallery", systemImage: "photo.on.rectangle")
                            .frame(maxWidth: .infinity).padding(14)
                            .background(Color(.systemGray6)).foregroundColor(.primary)
                            .cornerRadius(8)
                    }
                }
            }
        }
        .padding(.vertical, 4)
        .sheet(isPresented: $showImagePicker) {
            ImagePickerView(sourceType: sourceType) { image in
                capturedImage = image
                if let data = image.jpegData(compressionQuality: 0.8) {
                    onAnswer(questionId, data)
                }
            }
        }
    }
}

// ─── KYC Navigation Bar (iOS — sticky bottom) ────────────────────────────────

struct KYCNavigationBarView: View {
    let showBack: Bool
    let showSaveExit: Bool
    let nextLabel: String
    let isLoading: Bool
    let onBack: (() -> Void)?
    let onSaveExit: (() -> Void)?
    let onNext: () -> Void

    var body: some View {
        HStack {
            if showBack {
                Button(action: { onBack?() }) {
                    HStack(spacing: HiveSpacing.s1) {
                        Image(systemName: "chevron.left")
                        Text("Back").font(HiveTypography.bodyBase)
                    }
                    .foregroundColor(HiveColor.neutral.n700)
                }
            }
            if showSaveExit {
                Button("Save & Exit") { onSaveExit?() }
                    .font(HiveTypography.bodyBase)
                    .foregroundColor(HiveColor.brand.primary)
            }
            Spacer()
            Button(action: onNext) {
                if isLoading {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(tint: HiveColor.brand.white))
                } else {
                    Text(nextLabel)
                        .font(HiveComponent.button.font)
                }
            }
            .padding(.horizontal, HiveComponent.button.paddingH)
            .frame(height: HiveComponent.button.height)
            .background(isLoading ? HiveColor.neutral.n300 : HiveColor.brand.primary)
            .foregroundColor(HiveColor.brand.white)
            .cornerRadius(HiveComponent.button.borderRadius)
            .disabled(isLoading)
        }
        .padding(HiveSpacing.s4)
        .background(HiveColor.brand.white)
        .shadow(color: HiveShadows.navBar.color,
                radius: HiveShadows.navBar.radius,
                x: HiveShadows.navBar.x,
                y: HiveShadows.navBar.y)
    }
}

// ─── Placeholder for ImagePickerView (requires UIViewControllerRepresentable) ─

struct ImagePickerView: UIViewControllerRepresentable {
    let sourceType: UIImagePickerController.SourceType
    let onCapture: (UIImage) -> Void

    func makeUIViewController(context: Context) -> UIImagePickerController {
        let picker = UIImagePickerController()
        picker.sourceType = sourceType
        picker.delegate = context.coordinator
        return picker
    }
    func updateUIViewController(_ uiViewController: UIImagePickerController, context: Context) {}
    func makeCoordinator() -> Coordinator { Coordinator(onCapture: onCapture) }

    class Coordinator: NSObject, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
        let onCapture: (UIImage) -> Void
        init(onCapture: @escaping (UIImage) -> Void) { self.onCapture = onCapture }
        func imagePickerController(_ picker: UIImagePickerController,
                                    didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
            picker.dismiss(animated: true)
            if let image = info[.editedImage] as? UIImage ?? info[.originalImage] as? UIImage {
                onCapture(image)
            }
        }
        func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
            picker.dismiss(animated: true)
        }
    }
}

File: android-sdui/app/src/main/java/com/hsbc/sdui/components/kyc/KYCComponents.kt
package com.hsbc.sdui.components.kyc

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
// Hive tokens — defined locally since no external Hive library is present.
import com.hsbc.sdui.models.HiveBorderRadius
import com.hsbc.sdui.models.HiveColor
import com.hsbc.sdui.models.HiveComponent
import com.hsbc.sdui.models.HiveElevation
import com.hsbc.sdui.models.HiveSpacing
import com.hsbc.sdui.models.HiveTypography

private val HsbcRed   = HiveColor.Brand.Primary
private val HsbcGreen = HiveColor.Semantic.Success

// ─── KYC Text Input ───────────────────────────────────────────────────────────

data class KYCValidation(
    val required: Boolean = false,
    val minLength: Int? = null,
    val maxLength: Int? = null,
    val pattern: String? = null,
    val errorMessage: String? = null
)

@Composable
fun KYCTextInputComposable(
    questionId: String,
    label: String,
    placeholder: String = "",
    helpText: String? = null,
    validation: KYCValidation? = null,
    savedValue: String? = null,
    onAnswer: (String, String) -> Unit
) {
    var value by remember { mutableStateOf(savedValue ?: "") }
    var error by remember { mutableStateOf<String?>(null) }
    var touched by remember { mutableStateOf(false) }

    fun validate(v: String): String? {
        val vl = validation ?: return null
        if (vl.required && v.isBlank()) return vl.errorMessage ?: "This field is required"
        vl.minLength?.let { if (v.length < it) return "Minimum $it characters" }
        vl.pattern?.let {
            if (!Regex(it).matches(v)) return vl.errorMessage ?: "Invalid format"
        }
        return null
    }

    Column(modifier = Modifier.fillMaxWidth().padding(vertical = HiveSpacing.s1)) {
        Row {
            Text(label, style = HiveTypography.labelBase, color = HiveColor.Neutral.N700)
            if (validation?.required == true) {
                Text(" *", color = HiveColor.Semantic.Error, style = HiveTypography.labelBase)
            }
        }
        helpText?.let {
            Text(it, style = HiveTypography.caption, color = HiveColor.Neutral.N500,
                modifier = Modifier.padding(top = HiveSpacing.s1))
        }
        OutlinedTextField(
            value = value,
            onValueChange = { v ->
                value = v
                if (touched) error = validate(v)
            },
            placeholder = { Text(placeholder, color = HiveColor.Neutral.N400,
                                  style = HiveTypography.bodyBase) },
            isError = touched && error != null,
            textStyle = HiveTypography.bodyBase,
            modifier = Modifier.fillMaxWidth().padding(top = HiveSpacing.s1)
                .height(HiveComponent.Input.height),
            keyboardOptions = KeyboardOptions(
                capitalization = KeyboardCapitalization.Words,
                imeAction = ImeAction.Done
            ),
            keyboardActions = KeyboardActions(onDone = {
                touched = true
                error = validate(value)
                if (error == null) onAnswer(questionId, value)
            }),
            shape = RoundedCornerShape(HiveComponent.Input.borderRadius),
            colors = OutlinedTextFieldDefaults.colors(
                focusedBorderColor = HiveComponent.Input.focusColor,
                unfocusedBorderColor = HiveComponent.Input.borderColor,
                errorBorderColor = HiveComponent.Input.errorColor,
                focusedContainerColor = HiveComponent.Input.bgColor,
                unfocusedContainerColor = HiveComponent.Input.bgColor,
            )
        )
        if (touched && error != null) {
            Text(error!!, color = HiveColor.Semantic.Error, style = HiveTypography.caption,
                modifier = Modifier.padding(top = HiveSpacing.s1))
        }
    }
}

// ─── KYC Single Select — Modal Bottom Sheet ───────────────────────────────────

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun KYCSingleSelectComposable(
    questionId: String,
    label: String,
    options: List<Pair<String, String>>,   // value to label
    savedValue: String? = null,
    onAnswer: (String, String) -> Unit
) {
    var selected by remember { mutableStateOf(savedValue) }
    var showSheet by remember { mutableStateOf(false) }
    val selectedLabel = options.find { it.first == selected }?.second ?: "Select an option"

    Column(modifier = Modifier.fillMaxWidth().padding(vertical = HiveSpacing.s1)) {
        Text(label, style = HiveTypography.labelBase, color = HiveColor.Neutral.N700)
        Spacer(Modifier.height(HiveSpacing.s1))
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(HiveComponent.Input.borderRadius))
                .border(1.dp, HiveComponent.Input.borderColor,
                        RoundedCornerShape(HiveComponent.Input.borderRadius))
                .background(HiveComponent.Input.bgColor)
                .clickable { showSheet = true }
                .padding(horizontal = HiveComponent.Input.paddingH,
                         vertical = HiveComponent.Input.paddingV),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(selectedLabel,
                color = if (selected == null) HiveColor.Neutral.N400 else HiveColor.Neutral.N800,
                style = HiveTypography.bodyBase)
            Text("▼", color = HiveColor.Neutral.N500, style = HiveTypography.caption)
        }
    }

    if (showSheet) {
        ModalBottomSheet(onDismissRequest = { showSheet = false }) {
            Text(label, fontWeight = FontWeight.SemiBold, fontSize = 16.sp,
                modifier = Modifier.padding(horizontal = 16.dp, vertical = 12.dp))
            Divider()
            LazyColumn {
                items(options) { (value, optLabel) ->
                    ListItem(
                        headlineContent = { Text(optLabel) },
                        trailingContent = {
                            if (value == selected) Text("✓", color = HsbcGreen, fontWeight = FontWeight.Bold)
                        },
                        modifier = Modifier.clickable {
                            selected = value
                            showSheet = false
                            onAnswer(questionId, value)
                        }
                    )
                    Divider()
                }
            }
            Spacer(Modifier.height(32.dp))
        }
    }
}

// ─── KYC Progress Bar ─────────────────────────────────────────────────────────

@Composable
fun KYCProgressBarComposable(
    currentStep: Int,
    totalSteps: Int,
    sectionTitle: String,
    sectionProgress: String? = null
) {
    val progress = currentStep.toFloat() / totalSteps.toFloat()
    Column(modifier = Modifier.fillMaxWidth().padding(horizontal = HiveSpacing.s4, vertical = HiveSpacing.s2)) {
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            Text(sectionTitle, style = HiveTypography.caption, color = HiveColor.Neutral.N500)
            Text(sectionProgress ?: "Step $currentStep of $totalSteps",
                style = HiveTypography.caption, color = HiveColor.Neutral.N500)
        }
        Spacer(Modifier.height(HiveSpacing.s1))
        LinearProgressIndicator(
            progress = { progress },
            modifier = Modifier.fillMaxWidth()
                .height(HiveComponent.ProgressBar.height)
                .clip(RoundedCornerShape(HiveBorderRadius.full)),
            color = HiveComponent.ProgressBar.fillColor,
            trackColor = HiveComponent.ProgressBar.trackColor
        )
    }
}

// ─── KYC Navigation Bar ───────────────────────────────────────────────────────

@Composable
fun KYCNavigationBarComposable(
    showBack: Boolean,
    showSaveExit: Boolean,
    nextLabel: String,
    isLoading: Boolean = false,
    onBack: (() -> Unit)? = null,
    onSaveExit: (() -> Unit)? = null,
    onNext: () -> Unit
) {
    Surface(shadowElevation = HiveElevation.navBar) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(HiveSpacing.s4),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(horizontalArrangement = Arrangement.spacedBy(HiveSpacing.s2)) {
                if (showBack) {
                    OutlinedButton(onClick = { onBack?.invoke() },
                        border = BorderStroke(1.dp, HiveColor.Neutral.N300)) {
                        Text("← Back", style = HiveTypography.bodyBase,
                            color = HiveColor.Neutral.N700)
                    }
                }
                if (showSaveExit) {
                    TextButton(onClick = { onSaveExit?.invoke() }) {
                        Text("Save & Exit", style = HiveTypography.bodyBase,
                            color = HiveColor.Brand.Primary)
                    }
                }
            }
            Button(
                onClick = onNext,
                enabled = !isLoading,
                modifier = Modifier.height(HiveComponent.Button.height),
                contentPadding = PaddingValues(horizontal = HiveComponent.Button.paddingH),
                shape = RoundedCornerShape(HiveComponent.Button.borderRadius),
                colors = ButtonDefaults.buttonColors(
                    containerColor = HiveComponent.Button.bgColor,
                    disabledContainerColor = HiveColor.Neutral.N300
                )
            ) {
                if (isLoading) {
                    CircularProgressIndicator(color = HiveColor.Brand.White,
                        modifier = Modifier.size(18.dp), strokeWidth = 2.dp)
                } else {
                    Text(nextLabel, style = HiveComponent.Button.textStyle,
                        color = HiveComponent.Button.textColor)
                }
            }
        }
    }
}

// ─── KYC Declaration ─────────────────────────────────────────────────────────

@Composable
fun KYCDeclarationComposable(
    questionId: String,
    label: String,
    declarationText: String,
    onAnswer: (String, Boolean) -> Unit
) {
    var checked by remember { mutableStateOf(false) }
    Column(modifier = Modifier.fillMaxWidth().padding(16.dp)) {
        Text(label, fontWeight = FontWeight.SemiBold, fontSize = 16.sp)
        Spacer(Modifier.height(12.dp))
        Box(
            modifier = Modifier
                .fillMaxWidth().height(200.dp)
                .border(1.dp, Color.LightGray, RoundedCornerShape(8.dp))
                .padding(12.dp)
        ) {
            Text(declarationText, fontSize = 13.sp, color = Color.DarkGray,
                lineHeight = 20.sp)
        }
        Spacer(Modifier.height(12.dp))
        Row(verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.clickable {
                checked = !checked
                onAnswer(questionId, checked)
            }) {
            Checkbox(checked = checked, onCheckedChange = { v ->
                checked = v; onAnswer(questionId, v)
            }, colors = CheckboxDefaults.colors(checkedColor = HsbcRed))
            Spacer(Modifier.width(8.dp))
            Text("I have read and agree to the above declaration", fontSize = 14.sp)
        }
    }
}
